specs/approved/power-control-intermediary.rst

```diff
@@ -0,0 +1,741 @@
+provisioning tools, possibly requiring existing workflows to be completely
+refactored. There are many potential cases in which this would be very
+problematic to the end-user, and the alternative solution of providing the
+lessee with power credentials would be a huge security risk unacceptable to
+most administrators.
+
+
+Proposed change
+===============
+
+We will create a new REST API endpoint at ``{IRONIC URI}/redfish/v1`` that
+will act like a legitimate v1 Redfish endpoint to the end user. The
+functionality of this interface will provide strictly the essentials for
+Redfish-based provisioning tools to access power controls, as implementation
+of a full Redfish proxy would be beyond the scope of what we intend to do.
+(see `REST API Impact`_ below for more details). Minimally, it will allow for
+a user to identify themselves to Ironic and get/set the power state of any
+nodes that they have access to.
+
+These pseudo-Redfish API calls will be be translated into corresponding Ironic
+or Keystone API calls, and will allow for proper authentication as well as
+access to power controls. Provisioning capabilities will be available through
+Redfish's ``ComputerSystem`` interface and will provide equivalent
+functionality to the following Ironic CLI commands:
+
+* ``baremetal node list``
+* ``baremetal node show``
+* ``baremetal node power on``
+* ``baremetal node power off``
+* ``baremetal ndoe reboot``
+
+The workflow for authentication will be as follows:
+
+* User creates an application credential [2]_ for the project that has access
+  to the node they wish to access.
+* User uses the Redfish Sessions API to authenticate, specifying the
+  application credential UUID as the username, and the credential's secret
+  as the password.
+* User receives a token that grants them access to provisioning actions that
+  require authentication
+* Session is deleted; any further provisioning actions will require
+  re-authentication.
+
+This intermediary will abide by version 1.0.0 [6]_ of the Redfish
+specification for maximum backwards compatibility with existing tools.
+
+Alternatives
+------------
+
+The type of BMC interface emulation we're looking to implement here does
+already exist in sushy-tools [3]_ and VirtualBMC [4]_, which emulate
+Redfish and IPMI respectively. A previous spec was submitted by Tzu-Mainn
+Chen (tzumainn) which proposed the idea of a sushy-tools driver in Ironic to
+enable this functionality, but concerns about security, along with the
+potential value of this existing in Ironic proper have led to the proposal
+of this spec. [5]_
+
+As far as implementation goes, we currently plan on integrating this into the
+Ironic API itself, however it could also exist separately as its own separate
+WSGI service.
+
+Data model impact
+-----------------
+None.
+
+State Machine Impact
+--------------------
+None.
+
+REST API impact
+---------------
+
+The following endpoints will be added, available beginning with a new Bare
```

Comment:
Maybe emphasize that these endpoints mirror what Redfish provides?

Reply:
done + added citation to v1.0.0 of redfish spec + schema!